File: go/classes/reservation_system.go
// Reservation System — bookable resources with time intervals and conflict detection.
//
// Models airline seats and hotel rooms uniformly. Overlap is the core primitive:
// two reservations on the same resource collide iff their half-open intervals
// [start, end) overlap.
package classes

import (
	"errors"
	"fmt"
	"sort"
	"time"
)

// Resource is a specific bookable thing: "12A" or "Room 304".
type Resource struct {
	ID    string
	Kind  string // "seat", "room", "slot", ...
	Label string
}

// CustomerR is a person making reservations. Named CustomerR to avoid collision
// with movie_store.Customer in the same package.
type CustomerR struct {
	ID   string
	Name string
}

// ReservationStatus enumerates reservation lifecycle states.
type ReservationStatus string

const (
	ReservationBooked    ReservationStatus = "booked"
	ReservationCancelled ReservationStatus = "cancelled"
	ReservationCompleted ReservationStatus = "completed"
)

// Reservation is an interval claim on a resource by a customer.
type Reservation struct {
	ID         uint64
	ResourceID string
	CustomerID string
	Start      time.Time
	End        time.Time
	Status     ReservationStatus
	CreatedAt  time.Time
}

// IsActive reports whether the reservation is still booked.
func (r Reservation) IsActive() bool { return r.Status == ReservationBooked }

// Overlaps reports half-open interval overlap; touching boundaries do not conflict.
func (r Reservation) Overlaps(start, end time.Time) bool {
	return r.Start.Before(end) && start.Before(r.End)
}

// Reservation errors.
var (
	ErrUnknownResource    = errors.New("unknown resource")
	ErrUnknownReservation = errors.New("unknown reservation")
	ErrInvalidInterval    = errors.New("start must be before end")
	ErrConflict           = errors.New("reservation conflict")
	ErrAlreadyCancelled   = errors.New("reservation already cancelled")
)

// ReservationSystem allocates resources across time intervals.
type ReservationSystem struct {
	Resources    map[string]Resource
	Customers    map[string]CustomerR
	Reservations map[uint64]*Reservation
	nextID       uint64
}

// NewReservationSystem builds an empty system.
func NewReservationSystem() *ReservationSystem {
	return &ReservationSystem{
		Resources:    make(map[string]Resource),
		Customers:    make(map[string]CustomerR),
		Reservations: make(map[uint64]*Reservation),
		nextID:       1,
	}
}

// AddResource registers a resource.
func (s *ReservationSystem) AddResource(r Resource) error {
	if _, ok := s.Resources[r.ID]; ok {
		return fmt.Errorf("%w: resource %s", ErrDuplicate, r.ID)
	}
	s.Resources[r.ID] = r
	return nil
}

// AddCustomerR registers a customer.
func (s *ReservationSystem) AddCustomerR(c CustomerR) error {
	if _, ok := s.Customers[c.ID]; ok {
		return fmt.Errorf("%w: customer %s", ErrDuplicate, c.ID)
	}
	s.Customers[c.ID] = c
	return nil
}

// Conflicts returns active reservations on resourceID that overlap [start, end).
func (s *ReservationSystem) Conflicts(resourceID string, start, end time.Time) ([]*Reservation, error) {
	if _, ok := s.Resources[resourceID]; !ok {
		return nil, fmt.Errorf("%w: %s", ErrUnknownResource, resourceID)
	}
	if !start.Before(end) {
		return nil, ErrInvalidInterval
	}
	var out []*Reservation
	for _, r := range s.Reservations {
		if r.ResourceID == resourceID && r.IsActive() && r.Overlaps(start, end) {
			out = append(out, r)
		}
	}
	return out, nil
}

// Book creates a reservation, failing if the interval conflicts with existing bookings.
func (s *ReservationSystem) Book(resourceID, customerID string, start, end time.Time) (*Reservation, error) {
	if _, ok := s.Resources[resourceID]; !ok {
		return nil, fmt.Errorf("%w: %s", ErrUnknownResource, resourceID)
	}
	if _, ok := s.Customers[customerID]; !ok {
		return nil, fmt.Errorf("%w: customer %s", ErrUnknownCustomer, customerID)
	}
	if !start.Before(end) {
		return nil, ErrInvalidInterval
	}
	colliding, _ := s.Conflicts(resourceID, start, end)
	if len(colliding) > 0 {
		ids := make([]uint64, len(colliding))
		for i, c := range colliding {
			ids[i] = c.ID
		}
		return nil, fmt.Errorf("%w: %s overlaps %v", ErrConflict, resourceID, ids)
	}
	id := s.nextID
	s.nextID++
	r := &Reservation{
		ID:         id,
		ResourceID: resourceID,
		CustomerID: customerID,
		Start:      start,
		End:        end,
		Status:     ReservationBooked,
		CreatedAt:  time.Now(),
	}
	s.Reservations[id] = r
	return r, nil
}

// Cancel marks a reservation cancelled, freeing its interval.
func (s *ReservationSystem) Cancel(reservationID uint64) error {
	r, ok := s.Reservations[reservationID]
	if !ok {
		return fmt.Errorf("%w: %d", ErrUnknownReservation, reservationID)
	}
	if r.Status == ReservationCancelled {
		return fmt.Errorf("%w: %d", ErrAlreadyCancelled, reservationID)
	}
	r.Status = ReservationCancelled
	return nil
}

// Available returns resources (optionally filtered by kind) with no overlapping
// active reservation across [start, end).
func (s *ReservationSystem) Available(start, end time.Time, kind string) ([]Resource, error) {
	if !start.Before(end) {
		return nil, ErrInvalidInterval
	}
	var out []Resource
	for _, res := range s.Resources {
		if kind != "" && res.Kind != kind {
			continue
		}
		collides := false
		for _, r := range s.Reservations {
			if r.ResourceID == res.ID && r.IsActive() && r.Overlaps(start, end) {
				collides = true
				break
			}
		}
		if !collides {
			out = append(out, res)
		}
	}
	return out, nil
}

// Occupancy returns active reservations on a resource, sorted by start.
func (s *ReservationSystem) Occupancy(resourceID string) ([]*Reservation, error) {
	if _, ok := s.Resources[resourceID]; !ok {
		return nil, fmt.Errorf("%w: %s", ErrUnknownResource, resourceID)
	}
	var out []*Reservation
	for _, r := range s.Reservations {
		if r.ResourceID == resourceID && r.IsActive() {
			out = append(out, r)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Start.Before(out[j].Start) })
	return out, nil
}

// CustomerBookings returns every reservation a customer has made.
func (s *ReservationSystem) CustomerBookings(customerID string) ([]*Reservation, error) {
	if _, ok := s.Customers[customerID]; !ok {
		return nil, fmt.Errorf("%w: %s", ErrUnknownCustomer, customerID)
	}
	var out []*Reservation
	for _, r := range s.Reservations {
		if r.CustomerID == customerID {
			out = append(out, r)
		}
	}
	return out, nil
}

File: python/classes/reservation_system.py
"""Reservation System — bookable resources with time intervals and conflict detection.

Models airline seats and hotel rooms uniformly: a `Resource` is any bookable
slot, and a `Reservation` is an interval claim on one resource by one customer.
Overlap is the core primitive — two reservations on the same resource collide
iff their half-open intervals [start, end) overlap.
"""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Optional
import itertools


class ReservationStatus(Enum):
    BOOKED = "booked"
    CANCELLED = "cancelled"
    COMPLETED = "completed"


@dataclass
class Resource:
    """A specific bookable thing. `12A` or `Room 304` — not fungible."""
    id: str
    kind: str          # "seat", "room", "slot", ...
    label: str = ""
    attributes: dict[str, str] = field(default_factory=dict)


@dataclass
class Customer:
    id: str
    name: str


@dataclass
class Reservation:
    id: int
    resource_id: str
    customer_id: str
    start: datetime
    end: datetime
    status: ReservationStatus = ReservationStatus.BOOKED
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    def overlaps(self, start: datetime, end: datetime) -> bool:
        """Half-open [start, end) overlap — touching boundaries do not conflict."""
        return self.start < end and start < self.end

    @property
    def is_active(self) -> bool:
        return self.status == ReservationStatus.BOOKED


class ReservationError(Exception):
    """Base class for reservation errors."""


class UnknownEntity(ReservationError): ...
class Conflict(ReservationError): ...
class InvalidInterval(ReservationError): ...
class AlreadyCancelled(ReservationError): ...


class ReservationSystem:
    """Scheduled resource allocation. Availability is a negative search over conflicts."""

    def __init__(self) -> None:
        self.resources: dict[str, Resource] = {}
        self.customers: dict[str, Customer] = {}
        self.reservations: dict[int, Reservation] = {}
        self._ids = itertools.count(1)

    # --- registration ---
    def add_resource(self, resource: Resource) -> None:
        if resource.id in self.resources:
            raise ValueError(f"resource already registered: {resource.id}")
        self.resources[resource.id] = resource

    def add_customer(self, customer: Customer) -> None:
        if customer.id in self.customers:
            raise ValueError(f"customer already registered: {customer.id}")
        self.customers[customer.id] = customer

    # --- core: overlap and conflicts ---
    def conflicts(self, resource_id: str, start: datetime, end: datetime) -> list[Reservation]:
        """Active reservations on this resource whose interval overlaps [start, end)."""
        if resource_id not in self.resources:
            raise UnknownEntity(f"unknown resource: {resource_id}")
        if start >= end:
            raise InvalidInterval("start must be before end")
        return [r for r in self.reservations.values()
                if r.resource_id == resource_id and r.is_active and r.overlaps(start, end)]

    def book(self, resource_id: str, customer_id: str,
             start: datetime, end: datetime) -> Reservation:
        if resource_id not in self.resources:
            raise UnknownEntity(f"unknown resource: {resource_id}")
        if customer_id not in self.customers:
            raise UnknownEntity(f"unknown customer: {customer_id}")
        if start >= end:
            raise InvalidInterval("start must be before end")
        colliding = self.conflicts(resource_id, start, end)
        if colliding:
            raise Conflict(
                f"{resource_id} has {len(colliding)} overlapping reservation(s): "
                f"{[r.id for r in colliding]}")
        reservation = Reservation(
            id=next(self._ids),
            resource_id=resource_id,
            customer_id=customer_id,
            start=start,
            end=end,
        )
        self.reservations[reservation.id] = reservation
        return reservation

    def cancel(self, reservation_id: int) -> Reservation:
        if reservation_id not in self.reservations:
            raise UnknownEntity(f"unknown reservation: {reservation_id}")
        r = self.reservations[reservation_id]
        if r.status == ReservationStatus.CANCELLED:
            raise AlreadyCancelled(f"reservation {reservation_id} already cancelled")
        r.status = ReservationStatus.CANCELLED
        return r

    # --- queries ---
    def available(self, start: datetime, end: datetime,
                  kind: Optional[str] = None) -> list[Resource]:
        """Resources with no overlapping active reservation across [start, end)."""
        if start >= end:
            raise InvalidInterval("start must be before end")
        out: list[Resource] = []
        for res in self.resources.values():
            if kind is not None and res.kind != kind:
                continue
            if not self.conflicts(res.id, start, end):
                out.append(res)
        return out

    def occupancy(self, resource_id: str) -> list[Reservation]:
        """All active reservations on a resource, sorted by start."""
        if resource_id not in self.resources:
            raise UnknownEntity(f"unknown resource: {resource_id}")
        active = [r for r in self.reservations.values()
                  if r.resource_id == resource_id and r.is_active]
        return sorted(active, key=lambda r: r.start)

    def customer_bookings(self, customer_id: str) -> list[Reservation]:
        if customer_id not in self.customers:
            raise UnknownEntity(f"unknown customer: {customer_id}")
        return [r for r in self.reservations.values() if r.customer_id == customer_id]

    def summary(self) -> dict:
        active = [r for r in self.reservations.values() if r.is_active]
        return {
            "resources": len(self.resources),
            "customers": len(self.customers),
            "reservations_total": len(self.reservations),
            "reservations_active": len(active),
            "resources_booked": len({r.resource_id for r in active}),
        }


# --- demo ---
if __name__ == "__main__":
    from datetime import timedelta

    sys = ReservationSystem()
    sys.add_resource(Resource("12A", kind="seat", label="Window, row 12, first class"))
    sys.add_resource(Resource("12B", kind="seat", label="Aisle, row 12, first class"))
    sys.add_resource(Resource("304", kind="room", label="Hotel room 304, king bed"))
    sys.add_customer(Customer("C001", "Nathan"))
    sys.add_customer(Customer("C002", "Korrallan"))

    t = datetime(2026, 5, 1, 14, 0, tzinfo=timezone.utc)
    r1 = sys.book("12A", "C001", t, t + timedelta(hours=3))
    r2 = sys.book("304", "C001", t + timedelta(days=1), t + timedelta(days=4))

    try:
        sys.book("12A", "C002", t + timedelta(hours=1), t + timedelta(hours=4))
    except Conflict as e:
        print(f"conflict detected: {e}")

    r3 = sys.book("12B", "C002", t + timedelta(hours=1), t + timedelta(hours=4))
    free = sys.available(t, t + timedelta(hours=3), kind="seat")
    print(f"free seats during flight: {[r.id for r in free]}")
    print(sys.summary())
